sql/builders/aggregate.ts

```diff
@@ -0,0 +1,87 @@
+import {
+  build,
+  indent,
+  IOrdering,
+  ISelectBuilder,
+  ISelectOrBuilder,
+  ISelectSchema,
+  Select
+} from "df/sql/builders/select";
+
+export class AggregateBuilder<S extends ISelectSchema> implements ISelectBuilder<S> {
+  private selectedDimensions: ISelectSchema = {};
+  private selectedMetrics: ISelectSchema = {};
+  private whereClauses: string[] = [];
+  private selectedOrdering: IOrdering;
+  private selectedLimit: number;
+
+  constructor(private readonly from: ISelectOrBuilder<any>) {}
+
+  public metrics<MS extends ISelectSchema>(select: MS): AggregateBuilder<S & MS> {
+    this.selectedMetrics = { ...this.selectedMetrics, ...select };
+    return this;
+  }
+
+  public dimensions<DS extends ISelectSchema>(select: DS): AggregateBuilder<S & DS> {
+    this.selectedDimensions = { ...this.selectedDimensions, ...select };
+    return this;
+  }
+
+  public ordering(ordering: IOrdering) {
+    this.selectedOrdering = ordering;
+    return this;
+  }
+
+  public where(where: string) {
+    this.whereClauses.push(where);
+    return this;
+  }
+
+  public wheres(wheres: string[]) {
+    wheres.forEach(where => this.whereClauses.push(where));
+    return this;
+  }
+
+  public limit(limit: number) {
+    this.selectedLimit = limit;
+    return this;
+  }
+
+  public build() {
+    return Select.create<S>(
+      `select\n` +
+        (Object.keys(this.selectedDimensions).length + Object.keys(this.selectedMetrics).length ===
```

Comment:
some of the below confusion can be cleared up a lot if you pull out two boolean variables `selectingDimensions` and `selectingMetrics`